### PR TITLE
[routing-manager] add `OnLinkPrefixManager`

### PR DESCRIPTION
**Pull request by abtink**
This commit renames `LocalOnPreifx` class to `OnLinkPrefixManager` and
moves all the code related to managing of the on-link prefix to this
class, e.g., method `Evaluate()` to evaluate and make a decision
about on-link prefix (whether we need to advertise our local on-link
prefix or there is a favored discovered on-link prefix provided by
another router on infra link).

### CODE REVIEW

File: src/core/border_router/routing_manager.hpp [hunk 1]
```diff
@@ -632,13 +633,16 @@ class RoutingManager : public InstanceLocator
...
 
-        using ExpireTimer = TimerMilliIn<RoutingManager, &RoutingManager::HandleLocalOnLinkPrefixTimer>;
+        using ExpireTimer = TimerMilliIn<RoutingManager, &RoutingManager::HandleOnLinkPrefixManagerTimer>;
 
-        Ip6::Prefix mPrefix;
+        Ip6::Prefix mLocalPrefix;
+        Ip6::Prefix mDiscoveredPrefix;
         State       mState;
         TimeMilli   mExpireTime;
         Ip6::Prefix mOldPrefix;
```
Comment (wgtdkp) [inline src/core/border_router/routing_manager.hpp:648]: Nit: we probably want to make it clear that it is the old local on-link prefix, given that there is a new discovered on-link prefix now.
```suggestion
        Ip6::Prefix mOldLocalPrefix;
```

File: src/core/border_router/routing_manager.hpp [hunk 2]
```diff
@@ -632,13 +633,16 @@ class RoutingManager : public InstanceLocator
...
         void EnterAdvertisingState(void);
         void AppendCurPrefix(Ip6::Nd::RouterAdvertMessage &aRaMessage);
         void AppendOldPrefix(Ip6::Nd::RouterAdvertMessage &aRaMessage);
 
-        using ExpireTimer = TimerMilliIn<RoutingManager, &RoutingManager::HandleLocalOnLinkPrefixTimer>;
+        using ExpireTimer = TimerMilliIn<RoutingManager, &RoutingManager::HandleOnLinkPrefixManagerTimer>;
 
-        Ip6::Prefix mPrefix;
+        Ip6::Prefix mLocalPrefix;
+        Ip6::Prefix mDiscoveredPrefix;
```
Comment (wgtdkp) [inline src/core/border_router/routing_manager.hpp:645]: What about preserving the name "Favored Discovered Prefix"? The name worked for me :)
```suggestion
        Ip6::Prefix mFavoredDiscoveredPrefix;
```

**Review by wgtdkp — APPROVED** [review]
LGTM 👍 
Just a few nits!

File: src/core/border_router/routing_manager.hpp (hunk 2 above)
Comment (abtink) [reply src/core/border_router/routing_manager.hpp]: To be honest, I also like "favored discovered prefix" name :)...  
I know that I tend to use long names so sometimes try to actively edit myself (force myself to useshorter names)- Changed it back now to `mFavoredDiscoveredPrefix`.

File: src/core/border_router/routing_manager.hpp (hunk 1 above)
Comment (abtink) [reply src/core/border_router/routing_manager.hpp]: Changed in new push.

File: src/core/border_router/routing_manager.hpp (hunk 2 above)
Comment (wgtdkp) [reply src/core/border_router/routing_manager.hpp]: Thanks!

**Review by jwhui — APPROVED** [review]

**Pull request merged**